### PR TITLE
[PM-7253] Added mobile native docs

### PR DESCRIPTION
**Pull request by fedemkr**
## 🎟️ Tracking

[PM-7253](https://bitwarden.atlassian.net/browse/PM-7253)

## 📔 Objective

Update docs for the mobile native applications. Includes mobile native architecture and getting started with iOS.

**Notes:**
Getting started for Android will be handled in a different PR.

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation
  team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed
  issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

[PM-7253]: https://bitwarden.atlassian.net/browse/PM-7253?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: docs/getting-started/mobile/ios/index.md [hunk 1]
```diff
@@ -4,8 +4,92 @@ sidebar_position: 1
 
 # iOS
 
-This section documents how to get started with the iOS Bitwarden project in Swift.
+## Requirements
 
-## Repository
+1. [XCode](https://developer.apple.com/xcode/) (version 15.4)
```
Comment (KatherineInCode) [inline docs/getting-started/mobile/ios/index.md:9]: ⛏️ The official casing is "Xcode"

File: docs/getting-started/mobile/ios/index.md [hunk 2]
```diff
@@ -4,8 +4,92 @@ sidebar_position: 1
...
+   ```
+
+### Run the App
+
+1. Open the project in Xcode 15.4+.
+2. Run the app in the Simulator with the `Bitwarden` target.
+
+### Running Tests
+
+Due to slight snapshot test variations between iOS version, the test target requires running in an
```
Comment (KatherineInCode) [inline docs/getting-started/mobile/ios/index.md:62]:
```suggestion
Due to slight snapshot test variations between iOS versions, the test target requires running in an
```

**Review by KatherineInCode — DISMISSED** [review]

File: docs/architecture/mobile-clients/android/index.md [hunk 3]
```diff
@@ -4,16 +4,635 @@ sidebar_position: 1
...
-- **Target SDK**: 34
-- **Device Types Supported**: Phone and Tablet
-- **Orientations Supported**: Portrait and Landscape
+The responsibilities of the data layer are to manage the storage and retrieval of data from
+low-level sources (such as from the network, persistence, or Bitwarden SDK) and to expose them in a
+more ready-to-consume manner by the UI layer via "repository" and "manager" classes. The UI layer is
+then responsible for any final processing of this data for display in the UI as well for receiving
+events from the UI, updating the tracked state accordingly.
+
+## Data Layer
```
Comment (withinfocus) [inline docs/architecture/mobile-clients/android/index.md:31]:
```suggestion
## Data layer
```

File: docs/architecture/mobile-clients/android/index.md [hunk 4]
```diff
@@ -4,16 +4,635 @@ sidebar_position: 1
...
+
+The data layer is where all the UI-independent data is stored and retrieved. It consists of both raw
+data sources as well as higher-level "repository" and "manager" classes.
+
+Note that any functions exposed by a data layer class that must perform asynchronous work do so by
+exposing **suspending functions** that may run inside
+[coroutines](https://kotlinlang.org/docs/coroutines-guide.html) while any streaming sources of data
+are handled by exposing [Flows](https://kotlinlang.org/docs/flow.html).
+
+### Data Sources
```
Comment (withinfocus) [inline docs/architecture/mobile-clients/android/index.md:41]:
```suggestion
### Data sources
```

File: docs/architecture/mobile-clients/android/index.md [hunk 5]
```diff
@@ -4,16 +4,635 @@ sidebar_position: 1
...
+`Throwable`/`Exception` instances as part of "error" states should be avoided when possible.
+
+In some cases a source of data may be continuously observed and in these cases a repository may
+choose to expose a
+[StateFlow](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines.flow/-state-flow/)
+that emits data updates using the
+[DataState](https://github.com/bitwarden/android/blob/main/app/src/main/java/com/x8bit/bitwarden/data/platform/repository/model/DataState.kt)
+wrapper.
+
+### Note on Dependency Injection
```
Comment (withinfocus) [inline docs/architecture/mobile-clients/android/index.md:98]: 🎨 Make this an actual note or info block where appropriate.

File: docs/architecture/mobile-clients/android/index.md [hunk 6]
```diff
@@ -4,16 +4,635 @@ sidebar_position: 1
...
+[AuthDiskSource](https://github.com/bitwarden/android/blob/main/app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/disk/AuthDiskSource.kt)
+/
+[AuthDiskSourceImpl](https://github.com/bitwarden/android/blob/main/app/src/main/java/com/x8bit/bitwarden/data/auth/datasource/disk/AuthDiskSourceImpl.kt)).
+All `...Impl` classes are intended to be manually constructed while their associated interfaces are
+provided for dependency injection via a [Hilt Module](https://dagger.dev/hilt/modules.html) (ex:
+[PlatformNetworkModule](https://github.com/bitwarden/android/blob/main/app/src/main/java/com/x8bit/bitwarden/data/platform/datasource/network/di/PlatformNetworkModule.kt)).
+This prevents the `...Impl` classes from being injected by accident and allows the interfaces to be
+easily mocked/faked in tests.
+
+## UI Layer
```
Comment (withinfocus) [inline docs/architecture/mobile-clients/android/index.md:111]:
```suggestion
## UI layer
```

File: docs/architecture/mobile-clients/android/index.md [hunk 7]
```diff
@@ -4,16 +4,635 @@ sidebar_position: 1
...
+Navigation in the app is achieved via the
+[Compose Navigation](https://developer.android.com/develop/ui/compose/navigation) component. This
+involves specifying "destinations" within a
+[NavHost](https://developer.android.com/reference/androidx/navigation/NavHost) and using a
+[NavController](https://developer.android.com/reference/androidx/navigation/NavController) to
+trigger actual navigation events. The app uses a mix of state-based navigation at the highest level
+(in order to determine the overall app flow the user is in) and event-based navigation (to handle
+navigation within a flow based on user interactions).
+
+#### State-based Navigation
```
Comment (withinfocus) [inline docs/architecture/mobile-clients/android/index.md:526]:
```suggestion
#### State-based navigation
```

File: docs/architecture/mobile-clients/android/index.md [hunk 8]
```diff
@@ -4,16 +4,635 @@ sidebar_position: 1
...
+when logging out the current user, the data layer can decide the new state and the
+`RootNavViewModel` can decide where to go next, rather than having the screen that triggered the
+logout action choose the next navigation state (which would then need to be redundantly handled in
+multiple places in the app).
+
+State-based navigation should be limited to specific cases that affect the overall flow of the app
+and are driven by data emitted from the data layer. All other navigation should use event-based
+navigation.
+
+#### Event-based Navigation
```
Comment (withinfocus) [inline docs/architecture/mobile-clients/android/index.md:554]:
```suggestion
#### Event-based navigation
```

File: docs/architecture/mobile-clients/android/index.md [hunk 9]
```diff
@@ -4,16 +4,635 @@ sidebar_position: 1
...
+
+Event-based navigation is ultimately the result of some "action" sent to a VM, whether that comes
+directly from a user-interaction with the current screen or from some internal action of the VM sent
+as the result of some asynchronous call. The VM will send an "event" to the Compose layer, which
+will trigger a callback that will eventually make a call to a
+[NavController](https://developer.android.com/reference/androidx/navigation/NavController).
+
+#### Navigation Implementation
```
Comment (withinfocus) [inline docs/architecture/mobile-clients/android/index.md:562]:
```suggestion
#### Navigation implementation
```

File: docs/getting-started/mobile/ios/index.md [hunk 10]
```diff
@@ -4,8 +4,92 @@ sidebar_position: 1
 
 # iOS
```
Comment (withinfocus) [inline docs/getting-started/mobile/ios/index.md:5]: ❓ Is there not an Android version of this guide available?

**Review by withinfocus — COMMENTED** [review]
Big PR so I did a quick scan. I stopped commenting on the headers -- please get all of them.

File: docs/getting-started/mobile/ios/index.md (hunk 10 above)
Comment (fedemkr) [reply docs/getting-started/mobile/ios/index.md:5]: It's but we discussed it in Slack and it's going to be handled in a different PR by @patrick-bitwarden so we keep author attribution for the docs.

File: package.json [hunk 11]
```diff
@@ -45,11 +45,11 @@
...
     "@types/react": "18.3.3",
-    "cspell": "8.11.0",
-    "husky": "9.1.1",
-    "lint-staged": "15.2.7",
+    "cspell": "8.13.1",
+    "husky": "9.1.4",
+    "lint-staged": "15.2.8",
     "prettier": "3.3.3",
-    "typescript": "5.5.3"
+    "typescript": "5.5.4"
```
Comment (withinfocus) [inline package.json:52]: ♻️ These are usually good to separate from content changes unless connected.

**Review by withinfocus — APPROVED** [review]
Enough ⛏️s from me -- ship it!

File: package.json (hunk 11 above)
Comment (fedemkr) [reply package.json:52]: Interesting, I'll check on that given that it wasn't my intention to change it. Sorry I missed it.

**Review by withinfocus — DISMISSED** [review]

**Review by withinfocus — APPROVED** [review]
Just a rebase.

**Pull request merged**